File: lua/protein-translation/protein-translation.lua
local M = {}

local codon_to_protein = {}

local function add(protein, ...)
  for _, value in ipairs(table.pack(...)) do
    codon_to_protein[value] = protein
  end
end

add('Methionine', 'AUG')
add('Phenylalanine', 'UUU', 'UUC')
add('Leucine', 'UUA', 'UUG')
add('Serine', 'UCU', 'UCC', 'UCA', 'UCG')
add('Tyrosine', 'UAU', 'UAC')
add('Cysteine', 'UGU', 'UGC')
add('Tryptophan', 'UGG')
add('STOP', 'UAA', 'UAG', 'UGA')

function M.codon(codon) return codon_to_protein[codon] or error() end

function M.rna_strand(rna_strand)
  local result = {}

  for i = 1, #rna_strand, 3 do
    local codon = string.sub(rna_strand, i, i + 2)
    local protein = M.codon(codon)
    if protein == 'STOP' then break end

    table.insert(result, protein)
  end

  return result
end

return M

File: cpp/protein-translation/protein_translation.cpp
#include "protein_translation.h"
#include <cstdio>

namespace protein_translation {
std::vector<std::string> proteins(std::string_view input) {
  auto result = std::vector<std::string>();
  const size_t len = input.length();

  for (size_t i = 0; i < len; i += 3) {
    const auto codon = input.substr(i, 3);

    std::string acid;
    if (codon == "AUG") {
      acid = "Methionine";
    } else if (codon == "UUU" || codon == "UUC") {
      acid = "Phenylalanine";
    } else if (codon == "UUA" || codon == "UUG") {
      acid = "Leucine";
    } else if (codon == "UCU" || codon == "UCC" || codon == "UCA" ||
               codon == "UCG") {
      acid = "Serine";
    } else if (codon == "UAU" || codon == "UAC") {
      acid = "Tyrosine";
    } else if (codon == "UGU" || codon == "UGC") {
      acid = "Cysteine";
    } else if (codon == "UGG") {
      acid = "Tryptophan";
    } else if (codon == "UAA" || codon == "UAG" || codon == "UGA") {
      break;
    }

    result.push_back(acid);
  }

  return result;
}
} // namespace protein_translation
